datasources/src/main/java/org/opensearch/sql/datasources/storage/OpenSearchDataSourceMetadataStorage.java

```diff
@@ -217,13 +220,18 @@ private void createDataSourcesIndex() {
     }
   }
 
-  private List<DataSourceMetadata> searchInDataSourcesIndex(QueryBuilder query) {
+  private List<DataSourceMetadata> searchInDataSourcesIndex(
+      ClusterState state, QueryBuilder query) {
     SearchRequest searchRequest = new SearchRequest();
     searchRequest.indices(DATASOURCE_INDEX_NAME);
     SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
     searchSourceBuilder.query(query);
     searchSourceBuilder.size(DATASOURCE_QUERY_RESULT_SIZE);
     searchRequest.source(searchSourceBuilder);
+    if (state.isSegmentReplicationEnabled(DATASOURCE_INDEX_NAME)) {
```

Comment:
why not add parameter in search Request? e.g. consistentRead = true. Current approach expose internal implementation detail to client. In future, in case we add any feature breaking consistent read, all the client require consistent read need to make code change accordingly.

Reply:
+1 on Peng's comment.
Also are we doing this with all the plugins? 
Now, I am really not sure is it any worth of identifying the replication strategy and set the primary preference. What is the worst that could happen if we do it for all requests? It overwhelms the primary shard only if we cross certain limit? 
By any chance do we have numbers for that?

Reply:
Adding a consistency param would be cleaner here for clients/plugins and doesn't expose implementation detail or use internal apis.  However, this needs more vetting before being made api. 

We need to define the behavior here - does this simply route to primary or does it guarantee a strong read every time it is invoked? Because if its the latter we need to perform a refresh internally to guarantee even the primary is serving the latest writes irrespective of what refresh policy may have been set on the write path.  This concerns me because it changes refresh behavior where it may not be intended and have consequences.

I think what we really are asking for here is the former, which is to "read the last refreshed on data" whatever that may be at the time.

Reply:
We don't have any data on this @vamsi-amazon so I agree with leaving as is.  The downside of primary preference is the request will fail if the primary is unavailable and not route to replicas that are healthy.

To alleviate some concern, I suggest we use `_primary_first` which will attempt primary first and fallback to replicas only if the node is weighed away.

Reply:
@penghuo : Adding a new parameter may need more discussion and product feedback. 
@vamsi-amazon : No, we don't have data around primary shard limits i.e. max request that it can support before causing instability. 

Proceeding with @mch2 suggestion here.

Reply:
Updated. Request for review @penghuo @vamsi-amazon @mch2

Reply:
@penghuo @vamsi-amazon : Please have a look. Let know if you have any concerns. 

CC @mch2